src/validator/MatchValidator.cpp

```diff
@@ -0,0 +1,589 @@
+            if (!edgeInfos_[i].anonymous) {
+                columns->addColumn(makeColumn(*edgeInfos_[i].alias));
+            }
+            if (!nodeInfos_[i+1].anonymous) {
+                columns->addColumn(makeColumn(*nodeInfos_[i+1].alias));
+            }
+        }
+
+        if (columns->empty()) {
+            return Status::SemanticError("`RETURN *' not allowed if there is no alias");
+        }
+
+        ret->setColumns(columns);
+    }
+
+    // Check all referencing expressions are valid
+    std::vector<const Expression*> exprs;
+    exprs.reserve(ret->columns()->size());
+    for (auto *col : ret->columns()->columns()) {
+        exprs.push_back(col->expr());
+    }
+    NG_RETURN_IF_ERROR(validateAliases(exprs));
+
+    return Status::OK();
+}
+
+
+Status MatchValidator::validateAliases(const std::vector<const Expression*> &exprs) const {
+    static const std::unordered_set<Expression::Kind> kinds = {
+        Expression::Kind::kLabel,
+        Expression::Kind::kLabelAttribute
+    };
+
+    for (auto *expr : exprs) {
+        auto refExprs = ExpressionUtils::collectAll(expr, kinds);
+        if (refExprs.empty()) {
+            continue;
+        }
+        for (auto *refExpr : refExprs) {
+            auto kind = refExpr->kind();
+            const std::string *name = nullptr;
+            if (kind == Expression::Kind::kLabel) {
+                name = static_cast<const LabelExpression*>(refExpr)->name();
+            } else {
+                DCHECK(kind == Expression::Kind::kLabelAttribute);
+                name = static_cast<const LabelAttributeExpression*>(refExpr)->left()->name();
+            }
+            DCHECK(name != nullptr);
+            if (aliases_.count(*name) != 1) {
+                return Status::SemanticError("Alias used but not defined: `%s'", name->c_str());
+            }
+        }
+    }
+    return Status::OK();
+}
+
+
+Status MatchValidator::analyzeStartPoint() {
+    // TODO(dutor) Originate from either node or edge at any position
+    startFromNode_ = true;
+    startIndex_ = 0;
+    startExpr_ = nullptr;
+
+    auto &head = nodeInfos_[0];
+
+    if (head.label == nullptr) {
+        return Status::Error("Head node must have a label");
+    }
+    if (head.props == nullptr) {
+        return Status::Error("Head node must have a filter");
+    }
+
+    auto items = head.props->items();
+    DCHECK_EQ(items.size(), 1UL);
+    auto *indexFilter = new RelationalExpression(Expression::Kind::kRelEQ,
+            new TagPropertyExpression(
+                new std::string(*head.label),
+                new std::string(*items[0].first)),
+            items[0].second->clone().release());
+    saveObject(indexFilter);
+    scanInfo_.filter = indexFilter;
+    scanInfo_.schemaId = head.tid;
+
+    return Status::OK();
+}
+
+
+Expression* MatchValidator::createSubFilter(const std::string &alias,
+        const MapExpression *map) const {
+    DCHECK_NOTNULL(map);
```

Comment:
should not use `DCHECK_NOTNULL` to check variable nullable in one line. Replace it with `DCHECK(map != nullptr);` or `auto items = DCHECK_NOTNULL(map)->items();`

Reply:
thanks